helper constraints

FelipeAdachi opened the pull request:
## Description

### helper constraints:

- ~~is_complete~~ no_missing_values
- ~~is_unique~~
- is_in_range
- is_non_negative

Edit: is_unique has been removed until we figure out multi-metric constraints

### skip_missing

- refactor some helper constraints and metric constraint validation to concile `skip_missing` with `require_column_existence`

### tests

- tests for additional helper constraints and skip_missing logic

FelipeAdachi committed:
add helper constraints and update Constraints_Suite.ipynb

FelipeAdachi committed:
tests and precommit checks

jamie256 commented on python/tests/core/constraints/factories/test_count_metrics.py:
```diff
@@ -1,10 +1,21 @@
 from whylogs.core.constraints.factories import (
     count_below_number,
+    is_complete,
```
Consider renaming this something more explicit like "no_nulls" or "no_missing_values"

"is_complete" works but its broad enough one might expect that such a check is for completeness in terms of an expected row count or values in a range, or that there isn't any value missing in a set (e.g. the data contains values for all entries in a set).

FelipeAdachi replied:
done!

FelipeAdachi committed:
rename is_complete to no_missing_values

jamie256 commented on python/whylogs/core/constraints/factories/distribution_metrics.py:
```diff
@@ -60,6 +61,53 @@ def is_smaller(metric) -> bool:
+        upper bound of defined range
+    skip_missing: bool
+        If skip_missing is True, missing distribution metrics will make the check pass.
+        If False, the check will fail on missing metrics
```
suggest adding description to this comment:
"such as on an empty dataset"

FelipeAdachi replied:
done!

jamie256 commented on python/whylogs/core/constraints/factories/frequent_items.py:
```diff
@@ -1,8 +1,30 @@
 from ..metric_constraints import MetricConstraint, MetricsSelector
 
 
+def is_unique(column_name: str) -> MetricConstraint:
```
This is based on a probabilistic sketch, we need to name this something like "is_probably_unique" and it would be better perhaps to use cardinality estimate ~= total count.

FelipeAdachi replied:
based on previous conversations, I removed this until we get multi-metric constraints figured out

jamie256 reviewed (commented):
I think we need some naming tweaks to make sure these correctly describe what they are checking, but nice to have these helpers!

FelipeAdachi committed:
WIP: is_probabilistic_unique refactoring

FelipeAdachi committed:
WIP: is_probabilistic_unique

FelipeAdachi committed:
wip - is_probably_unique, column_pair_mean_a_less_or_equal_than_mean_b, skip_missing logic, testing

FelipeAdachi committed:
- remove is_probably_unique and column_pair_mean
- refactoring to concile skip_missing with require_column_existence
- more testing

FelipeAdachi committed:
Merge branch 'mainline' into dev/felipe/helper-constraints

murilommen commented on python/tests/core/constraints/factories/conftest.py:
```diff
@@ -18,3 +18,14 @@ def nan_builder():
+    df = pd.DataFrame(data)
+    profile_view = why.log(pandas=df).view()
+    empty_builder = ConstraintsBuilder(dataset_profile_view=profile_view)
+    return empty_builder
```
unless this is used in more than one test module, it should not be on the wider scope

murilommen commented on python/tests/core/constraints/factories/test_distribution_metrics.py:
```diff
@@ -1,13 +1,65 @@
+        "animal": ["cat", "hawk", "snake", "cat", "mosquito"],
+        "legs": [4, 2, 0, 4, 6],
+        "weight": [4.3, -1.8, 1.3, 4.1, 5.5e-6],
+        "null_column": [np.nan, np.nan, np.nan, np.nan, np.nan],
```
in case this changes for some reason

murilommen commented on python/whylogs/core/constraints/factories/distribution_metrics.py:
```diff
@@ -15,19 +15,14 @@ def greater_than_number(column_name: str, number: Union[float, int], skip_missin
     constraint = MetricConstraint(
         name=f"{column_name} greater than number {number}",
-        condition=is_greater,
+        condition=lambda x: x.min > number,
```
why has this changed back to lambda?

FelipeAdachi replied:
The same comment applies to all the other cases below

richard-rogers replied:
I think you could do `Require("min").greater_than(number)`.  Lambdas are often easier to code, but cannot be [de]serialized.

FelipeAdachi replied:
replaced lambda to Predicate in latest commit

murilommen commented on python/whylogs/core/constraints/factories/distribution_metrics.py:
```diff
@@ -73,19 +112,14 @@ def mean_between_range(column_name: str, lower: float, upper: float, skip_missin
     constraint = MetricConstraint(
         name=f"{column_name} mean between {lower} and {upper} (inclusive)",
-        condition=is_mean_between,
+        condition=lambda x: lower <= x.avg <= upper,
```
same, why lambda

murilommen commented on python/whylogs/core/constraints/factories/distribution_metrics.py:
```diff
@@ -103,19 +137,14 @@ def stddev_between_range(column_name: str, lower: float, upper: float, skip_miss
     constraint = MetricConstraint(
         name=f"{column_name} standard deviation between {lower} and {upper} (inclusive)",
-        condition=is_stddev_between_range,
+        condition=lambda x: lower <= x.stddev <= upper,
```
same

FelipeAdachi replied:
replaced lambda to Predicate in latest commit

FelipeAdachi commented on python/tests/core/constraints/factories/test_count_metrics.py:
```diff
@@ -1,10 +1,21 @@
+    builder.add_constraint(no_missing_values(column_name="animal"))
+    builder.add_constraint(no_missing_values(column_name="weight"))
+    constraint = builder.build()
+    assert constraint.validate() is False
```
constraints.validate() is False if any constraint in `constraints` fails. Since `weight` fails on the check, validate() returns false.
actually, `constraint` should be in plural here.

FelipeAdachi committed:
minor changes in testing

FelipeAdachi committed:
Merge branch 'dev/felipe/helper-constraints' of https://github.com/whylabs/whylogs into dev/felipe/helper-constraints

richard-rogers commented on python/tests/core/constraints/factories/test_distribution_metrics.py:
```diff
@@ -1,13 +1,65 @@
+    builder.add_constraint(is_non_negative(column_name="legs"))
+    builder.add_constraint(is_non_negative(column_name="animal", skip_missing=False))
+    builder.add_constraint(is_non_negative(column_name="null_column", skip_missing=True))
+    with pytest.raises(ValueError) as e:
```
```suggestion
    with pytest.raises(ValueError, match="non_existing_column was not found in set of this profile") as e:
```

Then you can drop the assert on line 52

richard-rogers commented on python/tests/core/constraints/factories/test_distribution_metrics.py:
```diff
@@ -1,13 +1,65 @@
+
+    constraint = builder.build()
+    report = constraint.generate_constraints_report()
+    assert report[0].name == "weight is non negative" and report[0].passed == 0 and report[0].failed == 1
```
parametrize?

FelipeAdachi replied:
done! thanks!

richard-rogers commented on python/whylogs/core/constraints/factories/distribution_metrics.py:
```diff
@@ -135,20 +164,18 @@ def quantile_between_range(column_name: str, quantile: float, lower: float, uppe
-            return result
-        else:
-            return True if skip_missing else False
+        quantile_value = metric.kll.value.get_quantile(quantile)
```
The `Predicate`s grab values to use in expressions from the metric summary, so I don't think they could handle this case. You're limited to the quantiles hard-coded in the summary.

FelipeAdachi replied:
maybe in the future, we could do serializable versions fetching from the hard-coded values in the summary.
Leaving this as is for now.

FelipeAdachi committed:
Merge branch 'mainline' into dev/felipe/helper-constraints

FelipeAdachi committed:
replacing lambdas to predicates and improving tests

FelipeAdachi committed:
Merge branch 'mainline' into dev/felipe/helper-constraints

jamie256 reviewed (approved):
LGTM!
